src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -2134,23 +2164,102 @@ void EmitAnchors(RegexNode node)
             // Emits the code to handle a multiple-character match.
             void EmitMultiChar(RegexNode node)
             {
-                // if (textSpanPos + node.Str.Length >= textSpan.Length) goto doneLabel;
-                // if (node.Str[0] != textSpan[textSpanPos]) goto doneLabel;
-                // if (node.Str[1] != textSpan[textSpanPos+1]) goto doneLabel;
-                // ...
-                EmitSpanLengthCheck(node.Str!.Length);
-                for (int i = 0; i < node.Str!.Length; i++)
+                bool caseInsensitive = IsCaseInsensitive(node);
+
+                // If the multi string's length exceeds the maximum length we want to unroll, instead generate a call to StartsWith.
+                // Each character that we unroll results in code generation that increases the size of both the IL and the resulting asm,
+                // and with a large enough string, that can cause significant overhead as well as even risk stack overflow due to
+                // having an obscenely long method.  Such long string lengths in a pattern are generally quite rare.  However, we also
+                // want to unroll for shorter strings, because the overhead of invoking StartsWith instead of doing a few simple
+                // inline comparisons is very measurable, especially if we're doing a culture-sensitive comparison and StartsWith
+                // accesses CultureInfo.CurrentCulture on each call.  We need to be cognizant not only of the cost if the whole
+                // string matches, but also the cost when the comparison fails early on, and thus we pay for the call overhead
+                // but don't reap the benefits of all the vectorization StartsWith can do.
+                const int MaxUnrollLength = 64;
+                if (node.Str!.Length > MaxUnrollLength)
                 {
+                    // if (!textSpan.Slice(textSpanPos).StartsWith("..."[, StringComparison.XxIgnoreCase])) goto doneLabel;
                     Ldloca(textSpanLocal);
-                    Ldc(textSpanPos + i);
-                    Call(s_spanGetItemMethod);
+                    Ldc(textSpanPos);
+                    Call(s_spanSliceIntMethod);
+                    Ldstr(node.Str);
+                    Call(s_stringAsSpanMethod);
+                    if (!caseInsensitive)
+                    {
+                        Call(s_spanStartsWith);
+                    }
+                    else
+                    {
+                        Ldc((int)(UseToLowerInvariant ? StringComparison.InvariantCultureIgnoreCase : StringComparison.CurrentCultureIgnoreCase));
```

Comment:
This won't produce same results as the compare one lower-cased character at a time algorithm used by the unrolled loop.

For example, `"o\u0308".StartsWith("o", StringComparison.InvariantCultureIgnoreCase)` returns false (https://github.com/dotnet/coreclr/pull/26759#issuecomment-533237086 has explanation why).

To match the behavior of the unrolled loop, I think you would need to have your own local implementation of StartsWith.